flink-streaming-connectors/flink-connector-elasticsearch2/src/main/java/org/apache/flink/streaming/connectors/elasticsearch2/ElasticsearchSink.java

```diff
@@ -186,22 +198,47 @@ public void beforeBulk(long executionId, BulkRequest request) {
 
 			@Override
 			public void afterBulk(long executionId, BulkRequest request, BulkResponse response) {
+				boolean allRequestsRepeatable = true;
 				if (response.hasFailures()) {
 					for (BulkItemResponse itemResp : response.getItems()) {
 						if (itemResp.isFailed()) {
-							LOG.error("Failed to index document in Elasticsearch: " + itemResp.getFailureMessage());
-							failureThrowable.compareAndSet(null, new RuntimeException(itemResp.getFailureMessage()));
+							// Check if index request can be retried
+							String failureMessageLowercase = itemResp.getFailureMessage().toLowerCase();
+							if (checkErrorAndRetryBulk && (
+									failureMessageLowercase.contains("timeout") || failureMessageLowercase.contains("timed out") // Generic timeout errors
+									|| failureMessageLowercase.contains("UnavailableShardsException".toLowerCase()) // Shard not available due to rebalancing or node down
+									|| (failureMessageLowercase.contains("data/write/bulk") && failureMessageLowercase.contains("bulk")) // Bulk index queue on node full
+									)
+								) {
+								LOG.debug("Retry bulk: {}", itemResp.getFailureMessage());
+								reAddBulkRequest(request);
```

Comment:
Are you sure that the `BulkRequest` is only added once even if it contains multiple failed `BulkItemResponse`s?